Source/WebDriver/Session.cpp

```diff
@@ -3144,4 +3145,183 @@ void Session::takeScreenshot(std::optional<String> elementID, std::optional<bool
     });
 }
 
+#if ENABLE(WEBDRIVER_BIDI)
+void Session::dispatchEvent(RefPtr<JSON::Object>&& message)
+{
+    static String automationPrefix = "Automation."_s;
+    auto method = message->getString("method"_s);
+    if (!method.startsWith(automationPrefix)) {
+        WTFLogAlways("Unknown event domain: %s", method.utf8().data());
+        return;
+    }
+
+    auto eventName = method.substring(automationPrefix.length());
+    if (!m_globalEventSet.contains(eventName)) {
+        WTFLogAlways("Ignoring unsubscribed event %s", eventName.utf8().data());
+        return;
+    }
+
+
+    if (eventName == "logEntryAdded"_s) {
+        doLogEntryAdded(WTFMove(message));
+        return;
+    }
+}
+
+void Session::doLogEntryAdded(RefPtr<JSON::Object>&& message)
+{
+    // https://w3c.github.io/webdriver-bidi/#event-log-entryAdded
+    auto params = message->getObject("params"_s);
+    auto method = params->getString("method"_s);
+
+    String level;
+    if (method == "error"_s || method == "assert"_s)
+        level = "error"_s;
+    else if (method == "debug"_s || method == "trace"_s)
+        level = "debug"_s;
+    else if (method == "warn"_s)
+        level = "warn"_s;
+    else
+        level = "info"_s;
+
+    // WebDriver uses the ECMA time format, which is the number of milliseconds since the Unix epoch.
+    // https://tc39.es/ecma262/#sec-time-values-and-time-range
+    RefPtr<JSON::Value> timestampValue;
+    if (auto timestampOpt = params->getDouble("timestamp"_s))
+        timestampValue = JSON::Value::create(round(*timestampOpt));
+    else
+        timestampValue = JSON::Value::create(0);
```

Comment:
you could rework this to use `timestampOpt.value_or(0)` instead of having the `else`